onnxscript/function_libs/torch_aten/graph_building.py

```diff
@@ -0,0 +1,446 @@
+"""Graph building functions for torchscript graph backend."""
+from __future__ import annotations
+
+import collections
+import typing
+import warnings
+from typing import Any, Optional, Sequence, Union
+
+import numpy as np
+import onnx
+import onnx.checker
+import onnx.defs
+import onnx.shape_inference
+import torch
+from beartype import beartype
+from torch.onnx import _type_utils
+
+import onnxscript
+from onnxscript import evaluator
+from onnxscript import tensor as onnxscript_tensor
+
+__all__ = [
+    "TorchScriptTensor",
+    "TorchScriptGraph",
+    "TorchScriptEvaluator",
+]
+
+
+ValidArgumentType = Union["TorchScriptTensor", str, int, float, bool]
+
+# TODO(justinchuby): Build a context manager to handle source information.
+
+
+class TorchScriptTensor(onnxscript_tensor.Tensor):
+    """A onnxscript tensor that wraps a torchscript Value."""
```

Comment:
Please add document (probably a design doc) describing how it should be used with other components (feel free to call me for details). The name is confusing; torchscript value can be non-tensor objects, so maybe we can call it TorchScriptValue.

Reply:
This is an onnxscript tensor. A little confusing indeed. We will see if we could name it better

Reply:
It may also be worth having a document/diagram illustrating the system and the flow of data through it. A key point is to pin down the representation-types used for different-types of values in different parts of the system (as much of the work is simply transforming between equivalent-representations used by different parts of the system).

Eg., I wonder if this could just be a wrapper around numpy arrays (which could serve as common intermediary format). Just to avoid too many layers of wrappers and inter-dependences.

Reply:
That’s a great point. Thinking about this again I wonder if it makes sense to only adapt numpy arrays (but not Python values) to Tensor in eager execution. So values flowing through systems in onnxscript all the way into Evaluator are either Tensor or Python primitives. 

This subclass of Tensor in particular needs to be symbolic and cannot have just a numpy array but needs to carry a reference to the node in the underlying graph (torch.Value) I think

Reply:
(Working on the graph)

Reply:
Sorry, I misunderstood. So, this is a placeholder, not a real tensor (is this what's referred to as FakeTensor elsewhere?)

It seems a bit fragile (to make Tensor.value None). Is it possible to get by without making this inherit from onnxscript Tensor?

Reply:
That is rely more on duck-typing, than actual inheritance-subtyping

Reply:
I considered making a `TensorProtocol` for this, but then the functions will then be annotated as `TensorProtocol[TReal]` etc., which is not exactly clean.

Alternatively we can rename `Tensor` to `ConcreteTensor`, then make `Tensor` the protocol, but then we need a tensor constructor e.g. `def tensor(nparray) -> ConcreteTensor` like pytorch does.

What do you think?

Reply:
> Sorry, I misunderstood. So, this is a placeholder, not a real tensor (is this what's referred to as FakeTensor elsewhere?)

That's accurate, although it _can_ carry a value when available too

Reply:
@gramalingam I also do want to inherit all the magic methods `__add__` etc. so we don't lose ops when uses use python symbols in a trace only function